File: setup-claude-server.js
import { homedir, platform } from 'os';
import { join } from 'path';
import { readFileSync, writeFileSync, existsSync, appendFileSync } from 'fs';
import { fileURLToPath } from 'url';
import { dirname } from 'path';
import { exec } from "node:child_process";

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

// Determine OS and set appropriate config path and command
const isWindows = platform() === 'win32';
const claudeConfigPath = isWindows
    ? join(process.env.APPDATA, 'Claude', 'claude_desktop_config.json')
    : join(homedir(), 'Library', 'Application Support', 'Claude', 'claude_desktop_config.json');

// Setup logging
const LOG_FILE = join(__dirname, 'setup.log');




function logToFile(message, isError = false) {
    const timestamp = new Date().toISOString();
    const logMessage = `${timestamp} - ${isError ? 'ERROR: ' : ''}${message}\n`;
    try {
        appendFileSync(LOG_FILE, logMessage);
        // For setup script, we'll still output to console but in JSON format
        const jsonOutput = {
            type: isError ? 'error' : 'info',
            timestamp,
            message
        };
        process.stdout.write(JSON.stringify(jsonOutput) + '\n');
    } catch (err) {
        // Last resort error handling
        process.stderr.write(JSON.stringify({
            type: 'error',
            timestamp: new Date().toISOString(),
            message: `Failed to write to log file: ${err.message}`
        }) + '\n');
    }
}

async function execAsync(command) {
    return new Promise((resolve, reject) => {
      // Use PowerShell on Windows for better Unicode support and consistency
      const actualCommand = isWindows
      ? `cmd.exe /c ${command}` 
      : command;
        
      exec(actualCommand, (error, stdout, stderr) => {
        if (error) {
          reject(error);
          return;
        }
        resolve({ stdout, stderr });
      });
    });
}

async function restartClaude() {
	try {
        const platform = process.platform
        switch (platform) {
            case "win32":
                // ignore errors on windows when claude is not running.
                // just silently kill the process
                try  {
                    await execAsync(
                        `taskkill /F /IM "Claude.exe"`,
                    )
                } catch {}
                break;
            case "darwin":
                await execAsync(
                    `killall "Claude"`,
                )
                break;
            case "linux":
                await execAsync(
                    `pkill -f "claude"`,
                )
                break;
        }
		await new Promise((resolve) => setTimeout(resolve, 3000))

		if (platform === "win32") {
            // it will never start claude
			// await execAsync(`start "" "Claude.exe"`)
		} else if (platform === "darwin") {
			await execAsync(`open -a "Claude"`)
		} else if (platform === "linux") {
			await execAsync(`claude`)
		}

		logToFile(`Claude has been restarted.`)
	} catch (error) {
		logToFile(`Failed to restart Claude: ${error}`, true)
	}
}

// Check if config file exists and create default if not
if (!existsSync(claudeConfigPath)) {
    logToFile(`Claude config file not found at: ${claudeConfigPath}`);
    logToFile('Creating default config file...');
    
    // Create the directory if it doesn't exist
    const configDir = dirname(claudeConfigPath);
    if (!existsSync(configDir)) {
        import('fs').then(fs => fs.mkdirSync(configDir, { recursive: true }));
    }
    
    // Create default config
    const defaultConfig = {
        "serverConfig": isWindows
            ? {
                "command": "cmd.exe",
                "args": ["/c"]
              }
            : {
                "command": "/bin/sh",
                "args": ["-c"]
              }
    };
    
    writeFileSync(claudeConfigPath, JSON.stringify(defaultConfig, null, 2));
    logToFile('Default config file created. Please update it with your Claude API credentials.');
}

try {
    // Read existing config
    const configData = readFileSync(claudeConfigPath, 'utf8');
    const config = JSON.parse(configData);

    // Prepare the new server config based on OS
    // Determine if running through npx or locally
    const isNpx = import.meta.url.endsWith('dist/setup-claude-server.js');

    const serverConfig = isNpx ? {
        "command": "npx",
        "args": [
            "@wonderwhy-er/desktop-commander"
        ]
    } : {
        "command": "node",
        "args": [
            join(__dirname, 'dist', 'index.js')
        ]
    };

    // Initialize mcpServers if it doesn't exist
    if (!config.mcpServers) {
        config.mcpServers = {};
    }
    
    // Check if the old "desktopCommander" exists and remove it
    if (config.mcpServers.desktopCommander) {
        logToFile('Found old "desktopCommander" installation. Removing it...');
        delete config.mcpServers.desktopCommander;
    }
    
    // Add or update the terminal server config with the proper name "desktop-commander"
    config.mcpServers["desktop-commander"] = serverConfig;

    // Write the updated config back
    writeFileSync(claudeConfigPath, JSON.stringify(config, null, 2), 'utf8');
    
    logToFile('Successfully added MCP server to Claude configuration!');
    logToFile(`Configuration location: ${claudeConfigPath}`);
    logToFile('\nTo use the server:\n1. Restart Claude if it\'s currently running\n2. The server will be available as "desktop-commander" in Claude\'s MCP server list');
    
    await restartClaude();

} catch (error) {
    logToFile(`Error updating Claude configuration: ${error}`, true);
    process.exit(1);
}

File: py/scripts/setup_claude_server.py
#!/usr/bin/env python3
import os
import sys
import json
import platform
import logging
import asyncio
from pathlib import Path


# Determine OS and set appropriate config path
def get_claude_config_path():
    """Get the path to Claude desktop config file based on OS."""
    system = platform.system()
    home_dir = os.path.expanduser("~")
    
    if system == "Windows":
        return os.path.join(os.environ.get("APPDATA", ""), "Claude", "claude_desktop_config.json")
    elif system == "Darwin":  # macOS
        return os.path.join(home_dir, "Library", "Application Support", "Claude", "claude_desktop_config.json")
    else:  # Linux and others
        return os.path.join(home_dir, ".config", "Claude", "claude_desktop_config.json")


# Setup logging
def setup_logging():
    """Set up logging for the setup script."""
    log_file = os.path.join(os.getcwd(), "setup.log")
    
    # Configure logging
    logging.basicConfig(
        level=logging.INFO,
        format="%(asctime)s - %(levelname)s - %(message)s",
        handlers=[
            logging.FileHandler(log_file),
            logging.StreamHandler(sys.stdout)
        ]
    )
    
    return logging.getLogger(__name__)


async def main():
    """Main setup function for Claude desktop integration."""
    logger = setup_logging()
    logger.info("Starting Claude desktop setup")
    
    # Get Claude config path
    claude_config_path = get_claude_config_path()
    logger.info(f"Claude config path: {claude_config_path}")
    
    # Check if config file exists and create default if not
    if not os.path.exists(claude_config_path):
        logger.info(f"Claude config file not found at: {claude_config_path}")
        logger.info("Creating default config file...")
        
        # Create parent directory if it doesn't exist
        os.makedirs(os.path.dirname(claude_config_path), exist_ok=True)
        
        # Create default config based on OS
        system = platform.system()
        if system == "Windows":
            default_config = {
                "serverConfig": {
                    "command": "cmd.exe",
                    "args": ["/c"]
                }
            }
        else:  # macOS/Linux
            default_config = {
                "serverConfig": {
                    "command": "/bin/sh",
                    "args": ["-c"]
                }
            }
        
        # Write default config
        with open(claude_config_path, "w") as f:
            json.dump(default_config, f, indent=2)
            
        logger.info("Default config file created")
    
    try:
        # Read existing config
        with open(claude_config_path, "r") as f:
            config = json.load(f)
        
        # Prepare the server config
        # Check if running from installed package or local directory
        script_path = os.path.abspath(sys.argv[0])
        is_installed = "site-packages" in script_path
        
        if is_installed:
            server_config = {
                "command": sys.executable,
                "args": ["-m", "desktop_commander.main"]
            }
        else:
            # Running from local directory
            server_config = {
                "command": sys.executable,
                "args": [os.path.join(os.getcwd(), "desktop_commander", "main.py")]
            }
        
        # Add or update the MCP server config
        if "mcpServers" not in config:
            config["mcpServers"] = {}
        
        config["mcpServers"]["desktopCommanderPy"] = server_config
        
        # Write the updated config back
        with open(claude_config_path, "w") as f:
            json.dump(config, f, indent=2)
        
        logger.info("Successfully added MCP server to Claude configuration!")
        logger.info(f"Configuration location: {claude_config_path}")
        logger.info(
            "\nTo use the server:\n"
            "1. Restart Claude if it's currently running\n"
            "2. The server will be available in Claude's MCP server list"
        )
        
    except Exception as e:
        logger.error(f"Error updating Claude configuration: {str(e)}")
        sys.exit(1)


if __name__ == "__main__":
    asyncio.run(main())